hoot-core/src/main/cpp/hoot/core/geometry/GeometryMerger.cpp

```diff
@@ -0,0 +1,209 @@
+/*
+ * This file is part of Hootenanny.
+ *
+ * Hootenanny is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * --------------------------------------------------------------------
+ *
+ * The following copyright notices are generated automatically. If you
+ * have a new notice to add, please use the format:
+ * " * @copyright Copyright ..."
+ * This will properly maintain the copyright information. DigitalGlobe
+ * copyrights will be updated automatically.
+ *
+ * @copyright Copyright (C) 2020 DigitalGlobe (http://www.digitalglobe.com/)
+ */
+
+#include "GeometryMerger.h"
+
+//  GEOS
+#include <geos/util/IllegalArgumentException.h>
+
+//  Hoot
+#include <hoot/core/geometry/GeometryUtils.h>
+#include <hoot/core/geometry/PolygonCompare.h>
+#include <hoot/core/util/ConfigOptions.h>
+#include <hoot/core/util/Log.h>
+#include <hoot/core/util/StringUtils.h>
+
+//  Qt
+#include <QString>
+#include <QStringList>
+
+//  Standard
+#include <thread>
+
+using namespace geos::geom;
+using namespace std;
+
+namespace hoot
+{
+
+GeometryMerger::GeometryMerger()
+  : _maxThreads(ConfigOptions().getCookieCutterAlphaShapeMaxThreads()),
+    _threadsWorking(true),
+    _finishedThreads(0),
+    _geometryCount(0),
+    _counter(0),
+    _updateInterval(ConfigOptions().getTaskStatusUpdateInterval())
+{
+}
+
+GeometryPtr GeometryMerger::mergeGeometries(std::vector<GeometryPtr> geometries, const geos::geom::Envelope& envelope)
```

Comment:
This function is a replacement for `GeometryUtils::mergeGeometries()` and kicks off a bunch of threads to do the merging in parallel.  This is allowed because of the way the algorithm was initially setup for reproducibility.  The `n`th and the `n+1`th elements are merged together and put into another vector.  Once the first vector is empty, the vectors are swapped and the process is repeated until there is only one geometry left.  The new algorithm does the same thing but the pairs are put on a work queue and distributed out to the thread pool.  Once a round is done, the main thread re-sorts them and creates another queue of pairs for the pool to work on.